Fix slow editor load on large projects (v2)

Hilderin opened the pull request:
- Fixes #94917

This PR includes the modifications of #95672 and the modifications in `core/io/file_access.cpp` from #93064.
This PR includes more modifications then #95672 and more possibilities of side effects. That's why I created a different PR.

The main problem while loading a project is the parsing and reading of all the .import and .md5 files. On a project of 60000 files, it means 120000 files. Worst was that each .import file was read 3 times in `EditorFileSystem::_test_for_reimport` because `ResourceFormatImporter::_get_path_and_type` was called internally 2 times at least. Once in `ResourceFormatImporter::are_import_settings_valid` and another time in `ResourceImporterLayeredTexture::are_import_settings_valid` while calling `ResourceFormatImporter::get_singleton()->get_resource_metadata(p_path)`

I did these modifications in `EditorFileSystem::_test_for_reimport`:
- Removed the `if (!FileAccess::exists(p_path + ".import"))`, usually, the file should exists and there's a `FileAccess::open` on the next line that returns an error if the file does not exists.
- The `metadata` is read directly instead of needing to call `ResourceFormatImporter::get_singleton()->get_resource_metadata(p_path)` which re-read the .import file
- `are_import_settings_valid` is called directly on the importer with the `metadata` preventing a call to `ResourceFormatImporter::get_singleton()->are_import_settings_valid` which re-read the .import file

I tested a project with all the images from https://github.com/PMDCollab/SpriteCollab/tree/master/sprite which contains 61332 png.

Before this PR:
- Loading the project in debug: 1m50sec
- Loading the project in 4.3 stable: 1m40sec

After this PR:
- Loading the project in debug: 50sec
- Loading the project in production build and optimize=speed lto=full: 30sec

Note: I got some variations when testing on the loading times. If I restart the editor quickly after another, the loading time is a bit faster. My guess is that some files are in the disk cache the second time around.

Hilderin committed:
Fix slow editor load on large projects (v2)

Saul2022 commented:
15 seconds to open the whole tps demo forge ( that icludes the time to enable the cronometer  and then going to godot to open the project (ON ANDROID s23+) , though i noticed that while super fast now scan after reloading got stuck at  89 sometimes or numbers like that  before reopening the scene. Also this is an artifact build so prob the end result could be faster ? ( not sure if artifacts are optimized..) 

Meanwhile on 4.3 stable it 25 seconds upon reloading or opening project.

smix8 commented:
Can confirm that this makes the rescan from e.g. renaming a file in a large project (100.000+ files) significantly faster, like 10000% faster.

Maybe unrelated to this PR but what is still super slow after this is the very first time editor start every day. For some reason it loads all the resource files again instead of just checking the meta data or md5. E.g. all the mesh assets in the project get reloaded from disk. Since the project has TB of mesh data this takes many minutes. Basically my old issue https://github.com/godotengine/godot/issues/47053

Hilderin commented:
> though i noticed that while super fast now scan after reloading got stuck at 89 sometimes or numbers like that before reopening the scene

That should be fixed by #93064, the issue is caused by the fact that the editor freezes after the file scan and the progress bar got stuck at the position where is was at the last frame.

Hilderin commented:
> Maybe unrelated to this PR but what is still super slow after this is the very first time editor start every day. For some reason it loads all the resource files again instead of just checking the meta data or md5. E.g. all the mesh assets in the project get reloaded from disk. Since the project has TB of mesh data this takes many minutes. Basically my old issue https://github.com/godotengine/godot/issues/47053

Unfortunately, I was never able to reproduce this problem on my side even if I tried on different OS and project :(

Saul2022 commented:
>, the issue is caused by the fact that the editor freezes after the file scan.

Alright i remembered when i tested it for other issue and that was fixed so it’s fine.

Aside from it you have any idea on how to improve losdong times in eg project with such massive data like the whole project i linked that had to wait 8 minutes just to scan every file and then it keeps not responding( and there wasn’t any scene that had to be reopened, so that’s because of the files.

Hilderin commented:
I'm not sure what linked project you are referring? I was not able to access the linked project from the original issue. As for the editor freeze, the PR #93064 should fix that. The first time a project is opened, it has to scan for files and after that it needs to analyze what to do with them. Currently, without #93064, there's no progress on this analysis process and with a lot of files and specially on Android, it could take a while. This PR also has some improvements on performance during this analysis phase but I would not expect to reduce dramatically the first loading time. I'm not really sure to understand correctly the problem without more details. Maybe create another issue with more details on your machine, your project, etc....?

Saul2022 commented:
> I'm not sure what linked project you are referring? I was not able to access the linked project from the original issue. 

I mean with the  https://github.com/PMDCollab/SpriteCollab/tree/master Whole  project downloading the zip file, 

> As for the editor freeze, the PR #93064 should fix that.

Sorry for confusion though i kind of confirmed it did with the popup dialog

> . Maybe create another issue with more details on your machine, your project, etc....?
Might do it when  this  and the dock v2 prs are merged to have the window, because it still kind of weird that  after scanning completely all files it takes so long to respond after the full scan when there’s no scene to do anything, just image files on the file system  , maybe that just related to the import system making adjustments for every single file.

Hilderin commented:
> Might do it when this and the dock v2 prs are merged to have the window, because it still kind of weird that after scanning completely all files it takes so long to respond after the full scan when there’s no scene to do anything, just image files on the file system , maybe that just related to the import system making adjustments for every single file.

Perfect. I also think it's because it's processing every file. It should be easier to pinpoint the problem once these PRs are merged. Thanks.

Saul2022 commented:
> Perfect. I also think it's because it's processing every file. It should be easier to pinpoint the problem once these PRs are merged. Thanks.

No problem i have seen many people complain for a long time on godot bad losdong times on big projects with lots of gbs, either seen people experience freezes,  editor quitting itself or crash after huge load times. As for processing every file , maybe as all this images uses the same sethings( because they all are pngs or jpegs) just  have one check to change  all the image properties at once without having to call it individually for them, not sure how it might work.

reduz commented:
This is really outstanding work, but there is something that is worrying me in general.
Technically, EditorFileSystem is designed so that those files (.import or anything else) are only parsed _on first editor load_ and then never again, since the file date and the information contained should be cached.

Basically, the expected behavior is that the second time you open the editor, the filesystem cache file is read, all files are simply checked for date if modified and if not nothing is done.

Still, I've seen reports or comments from users that even renaming a file is causing massive amount of work, so I have the feeling that this caching system has broken at some point and contributors kind of assume that "slow" is the way it should be expected to work.

I would really suggest investigating that the caching system is working to ensure we are not just band-aiding around a core bug  (even if,  of course, this optimization is still super welcome).

Saul2022 commented:
> , EditorFileSystem is designed so that those files (.import or anything else) are only parsed _on first editor load_ and then never again, since the file date and the information contained should be cached.

It kind of does no? Since on seconf load scan and opening stuff is faster than on the first. For example first time on trying to import sponza ivy glrf  ( over270 mb ) it  took 18:3 seconds to load the whole editor while on second time took 12-11 seconds after reloading the editor on the stable version, being the thing that holds like 6 seconds on both times the godot logo transition to the editor

Hilderin commented:
> Basically, the expected behavior is that the second time you open the editor, the filesystem cache file is read, all files are simply checked for date if modified and if not nothing is done.

That makes a lot more sense! Thanks for your feedback. I'll check that, I have a couple of ideas! Probably in another PR just to not mess with this one and to be careful.

Hilderin commented:
After some investigation, the main issue comes from calling `_test_for_reimport` on each asset file, which reads the `.import` and `md5` files, etc. This method was added in commit bb83c7d6b7c11c92c6c6bf3a151b3f0749d7c230 to fix issue #13135.

If I understand the original issue correctly, the logic should be:
- If the modification date changed:
  - If the hash changed:
    - Reimport

Currently, the logic is:
- If the modification date changed AND the hash has not changed:
  - Reimport

From what I understand, the original issue seems to point out that sometimes the last modification date could be different even if the file was not modified, which is probably the case in issue #47053, and we have confirmation of this on some Android devices (#94416). In that case, comparing the hashes could prevent a useless asset reimportation. The only missing part is a way to check if the content of the `.import` file is different to prevent issue #60730.

Is my understanding correct?

reduz commented:
@Hilderin Sorry I can't type much as I am on the phone. I think the idea is that the import file is only reparsed if the date changed. Then for imported assets, first date is checked and if it fails, then md5 is checked. Ultimately if both fail a reimport happens.

Hilderin commented:
I finally decided to implement all the modifications in this PR. Following @reduz suggestion, I investigated more deeply why we needed to read all these files on startup.

### Issues that I found:
- All the `.import` files were always read, even if the last modified date did not change, when the project setting `editor/import/reimport_missing_imported_files` is true, which is its default value.
- If `editor/import/reimport_missing_imported_files` is true, it was necessary to read the `.import` file to obtain the destination files of the import to validate if they still exist.
- In 2017, when the original MD5 was added, the method `_test_for_reimport` did not check the importer, importation settings, or MD5 on the first scan.
- Over the years, with each commit, additional steps were added, such as reading the MD5 file, reading the importer from the resource type to check the importer version, and validating the import settings during the first scan, which gradually slowed down the first scan.

### What I suggest in this PR:
- I added the MD5 of the `.import` file to the `filesystem_cache8`. This makes it easy to determine if the importation settings have changed and removes the dependence on the last modified date, which can be unreliable. This should fix issues #47053 and #94416 while still prevent issue #60730.
- I added the destination files to the `filesystem_cache8`, removing the need to read the `.import` file at all during editor startup.
- I divided the `_test_for_reimport` function into two distinct functions to clarify the process and prevent future errors. The first, `_is_test_for_reimport_needed`, is used only during the scan process, and I kept `_test_for_reimport` without the parameter `p_only_imported_files`, which is used when the last modification dates are different to verify more deeply if a reimportation is needed by validating the MD5, importer, importer settings, etc.

### The algorithm now works like this:
- Load the last `filesystem_cache8` if present.
- Scan process (first scan and scan for changes):
  - For each file, add the asset for an `ACTION_FILE_TEST_REIMPORT`:
    - If the last modification date of the asset file has changed
    - Or if the `.import` file is missing
    - Or if the last modification date of the `.import` file has changed
    - Or if `editor/import/reimport_missing_imported_files` is true and at least one of the destination files is missing
- Process each file action `ACTION_FILE_TEST_REIMPORT` (in `_update_scan_actions`):
  - Add the file for reimportation:
    - If the `.import` file is missing
    - Or if the MD5 for the `.import` file is different
    - Or if the importer name in the `.import` file does not exist
    - Or if the importer version is greater
    - Or if the importer settings are invalid
    - Or if the MD5 of the source asset file is different
    - Or if the MD5 of the destination files is different
- Reimport each file marked for reimportation.
- Save the `filesystem_cache8`.

### Time of the first scan:
- Before this PR: 1 min+
- After these modifications: 8-9 sec.

Saul2022 commented:
> I finally decided to implement all the modifications in this PR. Following @reduz suggestion, I investigated more deeply why we needed to read all these files on startup.

Wouldn't it be better to separate it on another  pr so it easier to merge this one first as is safer and then take the turbo speedup one for another. 


> ### Time of the scans:

> * Before this PR: 1 min+
> * After these modifications: 8-9 sec.

Can confirm , literal on a project with 5k files total ( https://github.com/user-attachments/files/16075644/test_base_mesh.zip ) the file scan is instant getting to 100 % just after the logo transition  ( which is already pretty fast now) the scene load the last what holding back, but that's for a future pr. Great job!

Hilderin commented:
> Wouldn't it be better to separate it on another pr so it easier to merge this one first as is safer and then take the turbo speedup one for another.

I'm not sure. I was thinking that it's easier to test everything at once and will prevent conflicts if the first PR needs changes. Also, we're not in a hurry; this problem has been present for a long time. I can gladly split this PR if the reviewers think it's better.

Repiteo commented:
I think this single-PR approach is preferable in this case; no real reason to have a standalone PR with just a bandaid solution only to immediately follow it with the *actual* fix. Either way, the changes affect core pretty heavily, so they'd be subject to thorough review/scrutiny regardless.

Saul2022 commented:
> I think this single-PR approach is preferable in this case; no real reason to have a standalone PR with just a bandaid solution only to immediately follow it with the _actual_ fix. 

Alright then , i am just excited to have it merged and possibly combine witht he dock pr  to speed even further the scan speed  and editor feedback. 

Also for  pushing the file system scan further i increased tps files to over 39.87gb and  over 35k files  ( .godot included) and now whole load is 33.09 seconds( decent jump when with 2.5 gb it was 13 secs) Still pushing well for a phone, but curious on why smix gdscript takes less than 2 seconds to scan his project with tons of gbs..

Hilderin commented:
Rebased to fix conflict with master.

akien-mga commented on editor/editor_file_system.cpp:
```diff
@@ -369,6 +384,11 @@ void EditorFileSystem::_scan_filesystem() {
 					fc.script_class_name = split[7].get_slice("<>", 0);
 					fc.script_class_extends = split[7].get_slice("<>", 1);
 					fc.script_class_icon_path = split[7].get_slice("<>", 2);
+					fc.import_md5 = split[7].get_slice("<>", 3);
+					String dest_paths = split[7].get_slice("<>", 4);
```
With so many slices for the same separator, I wonder if it wouldn't be faster to actually use `split()` and then index the list.

Hilderin replied:
After some testing on a project with more then 60000 files, it's virtually the same performance. It took around 900ms to load the file with the get_slice and the split.
And the code is easier to read with the get_slice:
```cpp
Vector<String> other_infos = split[7].split("<>");
int size_other_infos = other_infos.size();
if (size_other_infos > 0) {
  fc.script_class_name = other_infos[0];
}
if (size_other_infos > 1) {
   fc.script_class_extends = other_infos[1];
}
if (size_other_infos > 2) {
  fc.script_class_icon_path = other_infos[2];
}
if (size_other_infos > 3) {
  fc.import_md5 = other_infos[3];
}
if (size_other_infos > 4) {
  String dest_paths = other_infos[4];
  if (!dest_paths.is_empty()) {
    fc.import_dest_paths = dest_paths.split("<*>");
  }
}
```
So, if it's ok with, you, I suggest to keep the get_slice?

akien-mga replied:
Well I would have done:
```cpp
Vector<String> other_infos = split[7].split("<>");
ERR_FAIL_COND(other_infos.size() < 5);
fc.script_class_name = other_infos[0];
fc.script_class_extends = other_infos[1];
fc.script_class_icon_path = other_infos[2];
fc.import_md5 = other_infos[3];
String dest_paths = other_infos[4];
if (!dest_paths.is_empty()) {
    fc.import_dest_paths = dest_paths.split("<*>");
}
```
which seems pretty readable to me.

But if there's no performance difference, either way is fine.

Hilderin commented:
Rebased to fix merge conflicts.

H2xDev commented:
I confirm as well. The fix is working good on real project with 4.5gb of assets (about 15k files). Thank you!

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

Benchmark on the TPS demo:

```
Benchmark 1: bin/godot.linuxbsd.editor.x86_64 ~/Documents/Godot/tps-demo/project.godot --quit      <-------- This PR
  Time (mean ± σ):      3.464 s ±  0.534 s    [User: 2.220 s, System: 0.723 s]
  Range (min … max):    2.971 s …  4.253 s    10 runs
 
Benchmark 2: bin/godot.linuxbsd.editor.x86_64.master ~/Documents/Godot/tps-demo/project.godot --quit
  Time (mean ± σ):      3.943 s ±  0.224 s    [User: 2.198 s, System: 0.731 s]
  Range (min … max):    3.310 s …  4.045 s    10 runs

Summary
  bin/godot.linuxbsd.editor.x86_64 ~/Documents/Godot/tps-demo/project.godot --quit ran
    1.14 ± 0.19 times faster than bin/godot.linuxbsd.editor.x86_64.master ~/Documents/Godot/tps-demo/project.godot --quit
```

AThousandShips commented on editor/editor_file_system.cpp:
```diff
@@ -591,47 +640,98 @@ bool EditorFileSystem::_test_for_reimport(const String &p_path, bool p_only_impo
+			} else if (assign == "files") {
+				Array fa = value;
+				for (int i = 0; i < fa.size(); i++) {
+					dest_paths.push_back(fa[i]);
```
Can use for-each here

Hilderin replied:
Done!

H2xDev commented:
Could it be cherrypicked for 4.3.x?

KoBeWi commented on editor/import/resource_importer_texture.cpp:
```diff
@@ -740,16 +740,14 @@ String ResourceImporterTexture::get_import_settings_string() const {
 		}
 
-		String editor_meta_path = imported_path.replace(".editor.ctex", ".editor.meta");
+		String editor_meta_path = imported_path.replace(".editor.ctex", ".editor.p_meta");
```
Looks like you replaced too much.

Hilderin replied:
Indeed, Done!

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -1625,7 +1723,7 @@ void EditorFileSystem::_save_filesystem_cache(EditorFileSystemDirectory *p_dir,
 		cache_string.append(itos(file_info->import_valid));
 		cache_string.append(file_info->import_group_file);
-		cache_string.append(String("<>").join({ file_info->script_class_name, file_info->script_class_extends, file_info->script_class_icon_path }));
+		cache_string.append(String("<>").join({ file_info->script_class_name, file_info->script_class_extends, file_info->script_class_icon_path, file_info->import_md5, String("<*>").join(file_info->import_dest_paths) }));
```
This shouldn't be added here. You could just append at the end.

Hilderin replied:
I added the field there because the last field is special, it can contains the `::` separator. I don't think it will be possible to add more fields without breaking compatibility.
Check the line 367 in editor_file_system.cpp
```cpp
// The last section (deps) may contain the same splitter, so limit the maxsplit to 8 to get the complete deps.
Vector<String> split = l.split("::", true, 8);
```

KoBeWi replied:
Maybe add it before last field? (will require some code to handle old cache)
In worst case we can bump cache file, so it gets invalidated.
It's just weird to have a list of script cache stuff and then push some random checksums. The cache is already quite messy.

Though if it requires invalidating anyway, it could be done in the future™ I guess.

KoBeWi replied:
Yeah it's dumb that dependencies are just dumped at the end. It makes it impossible to properly add new cache entries. This should be changed, but not in this PR.

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -463,12 +483,33 @@ void EditorFileSystem::_thread_func(void *_userdata) {
-	if (!reimport_on_missing_imported_files && p_only_imported_files) {
-		return false;
+bool EditorFileSystem::_is_test_for_reimport_needed(const String &p_path, uint64_t p_last_modification_time, uint64_t p_modification_time, uint64_t p_last_import_modification_time, uint64_t p_import_modification_time, const Vector<String> &p_import_dest_paths) {
+	// The idea here is to trust the cache if the last modification times of the file and the .import file
+	// are the same. That means the files did not change since the last importation and that the files
```
This sounds like it checks whether modification time of file is the same as modification time of .import. Needs slight rewording.

Hilderin replied:
Indeed, it was not clear at all, I think it's better now.

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -591,47 +640,98 @@ bool EditorFileSystem::_test_for_reimport(const String &p_path, bool p_only_impo
-	//imported files are gone, reimport
-	for (const String &E : to_check) {
-		if (!FileAccess::exists(E)) {
+	//check source md5 matching
```
```suggestion
	// Check source md5 matching.
```
Same with other comments.

Hilderin replied:
Done!

KoBeWi replied:
You only fixed this one comment ;)

Hilderin replied:
You are right! Done!

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -591,47 +640,98 @@ bool EditorFileSystem::_test_for_reimport(const String &p_path, bool p_only_impo
+		return true;
+	}
+
+	if (dest_files.size() && !dest_md5.is_empty()) {
```
```suggestion
	if (!dest_files.is_empty() && !dest_md5.is_empty()) {
```

Hilderin replied:
Done!

KoBeWi reviewed (approved):
I get 4s speedup for my project load, but most of the startup time is outside scanning the filesystem. The scanning itself is noticeably faster though.

Did some testing and didn't find bugs.

akien-mga commented:
Thanks! Awesome work :tada:

fgheorghe commented:
Lurker, but my project has 100k assets and this will greatly improve my workflow. Thank you for this! Awesome work indeed!!

Saul2022 commented:
Victory!!!!!!!! Retty helpful pr and fast.

hayahane commented:
It looks like the new `_is_test_for_reimport_needed` doesn't consider assets that has an editor variant (Introduced in https://github.com/godotengine/godot/pull/64938). `ResourceFormatImporter::get_singleton()->are_import_settings_valid(path)` are not called when editor theme settings changed.
